File: frontend/src/lib/challenges.ts
import { api } from "./api";

export type ChallengeKind = "movie_count" | "episode_count" | "genre_count";
export type ChallengeStatus = "upcoming" | "active" | "ended";

export type Challenge = {
  id: string;
  title: string;
  description: string | null;
  kind: ChallengeKind;
  genre_name: string | null;
  target_count: number;
  starts_at: string;
  ends_at: string;
  status: ChallengeStatus;
  progress: number;
  earned: boolean;
};

export type ChallengeCreate = {
  title: string;
  description?: string | null;
  kind: ChallengeKind;
  genre_name?: string | null;
  target_count: number;
  starts_at: string;
  ends_at: string;
};

export type LeaderboardUser = { id: string; username: string; avatar_url: string | null };
export type LeaderboardEntry = { user: LeaderboardUser; progress: number; earned: boolean; is_viewer: boolean };
export type LeaderboardResponse = { challenge: Challenge; entries: LeaderboardEntry[] };

export async function getChallenges(): Promise<Challenge[]> {
  const { data } = await api.get<{ results: Challenge[] }>("/challenges");
  return data.results;
}

export async function createChallenge(payload: ChallengeCreate): Promise<Challenge> {
  const { data } = await api.post<Challenge>("/challenges", payload);
  return data;
}

export async function deleteChallenge(challengeId: string): Promise<void> {
  await api.delete(`/challenges/${challengeId}`);
}

export async function getLeaderboard(challengeId: string): Promise<LeaderboardResponse> {
  const { data } = await api.get<LeaderboardResponse>(`/challenges/${challengeId}/leaderboard`);
  return data;
}

File: backend/app/schemas/challenges.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class ChallengeCreate(BaseModel):
    title: str = Field(min_length=1, max_length=100)
    description: str | None = Field(default=None, max_length=500)
    kind: Literal["movie_count", "episode_count", "genre_count"]
    genre_name: str | None = None
    target_count: int = Field(ge=1, le=1000)
    starts_at: datetime
    ends_at: datetime


class ChallengeOut(BaseModel):
    id: str
    title: str
    description: str | None
    kind: Literal["movie_count", "episode_count", "genre_count"]
    genre_name: str | None
    target_count: int
    starts_at: datetime
    ends_at: datetime
    status: Literal["upcoming", "active", "ended"]
    progress: int
    earned: bool


class ChallengeListResponse(BaseModel):
    results: list[ChallengeOut]


class LeaderboardUser(BaseModel):
    id: str
    username: str
    avatar_url: str | None


class LeaderboardEntry(BaseModel):
    user: LeaderboardUser
    progress: int
    earned: bool
    is_viewer: bool


class LeaderboardResponse(BaseModel):
    challenge: ChallengeOut
    entries: list[LeaderboardEntry]
